lib/livebook_web/controllers/session_controller.ex

```diff
@@ -127,7 +127,14 @@ defmodule LivebookWeb.SessionController do
       conn
       |> cache_permanently()
       |> put_status(:moved_permanently)
-      |> redirect(to: ~p"/public/sessions/node/#{node_id}/assets/#{hash}/#{file_parts}")
+      |> redirect(
+        to:
+          ~p"/public/sessions/node/#{node_id}/assets/#{hash}/#{file_parts}"
+          |> String.replace(
+            Livebook.Config.base_url_path(),
+            Livebook.Config.public_base_url_path()
+          )
```

Comment:
Replacing in the whole string may break in very specific cases. Let's define a function:

```elixir
def public_path(path) do
  base_url_path = Livebook.Config.base_url_path()
  public_base_url_path = Livebook.Config.public_base_url_path()
  ^base_url_path <> rest = url
  public_base_url_path <> rest
end
```

We can put it in `LivebookWeb` and import in `verified_routes`.

Reply:
I'm sorry, I don't know how to make `verified_routes` use the newly defined function.

I've tried removing the `String.replace()` call from the highlighted code, and adding this to `LivebookWeb`, but it doesn't seem to be doing anything
```elixir
def public_path(path) do
    base_url_path = Livebook.Config.base_url_path()
    public_base_url_path = Livebook.Config.public_base_url_path()
    ^base_url_path <> rest = path
    public_base_url_path <> rest
  end

  def verified_routes do
    quote do
      use Phoenix.VerifiedRoutes,
        endpoint: LivebookWeb.Endpoint,
        router: LivebookWeb.Router,
        statics: LivebookWeb.static_paths()

      # We don't know the hostname Livebook runs on, so we don't use
      # absolute URL helpers
      import Phoenix.VerifiedRoutes, only: :sigils
      import LivebookWeb, only: [public_path: 1]
    end
  end
  ```
  Could you explain what I need to do, please?

Reply:
@elepedus oh, this looks good, we still need to call it explicitly, but at least it's more convenient. So in the highlighted code, do this:

```
public_path(~p"/public/sessions/node/#{node_id}/assets/#{hash}/#{file_parts}")
```

Reply:
Awesome -- I've done that :) 
Out of curiosity, would there be a way to integrate that new function with `sigil_p`, so we don't need to call it explicitly?

Reply:
Aaactually, we can :)

```elixir
def verified_routes do
  quote do
    use Phoenix.VerifiedRoutes,
      endpoint: LivebookWeb.Endpoint,
      router: LivebookWeb.Router,
      statics: LivebookWeb.static_paths()

    # We don't know the hostname Livebook runs on, so we don't use
    # absolute URL helpers. We don't import sigil_p either, because
    # we override it.
    import Phoenix.VerifiedRoutes, only: []

    import LivebookWeb, only: [sigil_p: 2]
  end
end

# ...

# Overrides

require Phoenix.VerifiedRoutes

defmacro sigil_p({:<<>>, _meta, ["/public/" <> _ | _]} = route, extra) do
  # We allow configuring a base path for all routes and we configure
  # Phoenix to use it. However, we have an additional configuration
  # for base path applying only to /public. We use a custom sigil_p
  # to insert the base path if needed.
  quote do
    path = Phoenix.VerifiedRoutes.sigil_p(unquote(route), unquote(extra))
    LivebookWeb.__rewrite_public_base_path__(path)
  end
end

defmacro sigil_p(route, extra) do
  quote do
    Phoenix.VerifiedRoutes.sigil_p(unquote(route), unquote(extra))
  end
end

def __rewrite_public_base_path__(path) do
  base_url_path = Livebook.Config.base_url_path()
  public_base_url_path = Livebook.Config.public_base_url_path()
  ^base_url_path <> rest = path
  public_base_url_path <> rest
end
```

@josevalim wdyt?

Reply:
That looks fine to me!

Reply:
Oh, that's really cool! 😎 

I'm curious, though, why do we still need `import Phoenix.VerifiedRoutes, only: []`?

Reading naively, it looks like we're saying "import nothing from Phoenix.VerifiedRoutes", which should be equivalent to deleting the line entirely?

Reply:
`use Phoenix.VerifiedRoutes` imports everything ([ref](https://github.com/phoenixframework/phoenix/blob/25e891f172045aa0391d30bfd794111235041159/lib/phoenix/verified_routes.ex#L128)), so we need to unimport it :)

Reply:
Oh! I see! Thanks! :)